File: core/cyp/backtest/archive.py
"""OHLCV 归档：从 CEX 分页拉取历史 K 线 → PostgreSQL/TimescaleDB 增量缓存，供回测使用真实历史。

- 公共行情无需 API Key（CexVenue 只读即可）。
- 缓存按 (venue, symbol, timeframe, ts) 去重，重复拉取只补缺口。
- ohlcv 表为 TimescaleDB hypertable（按 ts 分区），为后续大规模历史行情做准备。
- 离线可测：venue 只需实现 fetch_ohlcv(symbol, timeframe, limit, since)。
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from decimal import Decimal

import asyncpg

from cyp.contracts import Candle
from cyp.memory.store import default_db_url

_TF_MINUTES = {"1m": 1, "5m": 5, "15m": 15, "30m": 30, "1h": 60, "4h": 240, "1d": 1440}

_SCHEMA_STMTS = (
    "CREATE EXTENSION IF NOT EXISTS timescaledb",
    """
    CREATE TABLE IF NOT EXISTS ohlcv (
        venue TEXT NOT NULL,
        symbol TEXT NOT NULL,
        timeframe TEXT NOT NULL,
        ts TIMESTAMPTZ NOT NULL,
        open NUMERIC NOT NULL,
        high NUMERIC NOT NULL,
        low NUMERIC NOT NULL,
        close NUMERIC NOT NULL,
        volume NUMERIC NOT NULL,
        PRIMARY KEY (venue, symbol, timeframe, ts)
    )
    """,
    "SELECT create_hypertable('ohlcv', 'ts', if_not_exists => TRUE, migrate_data => TRUE)",
)

_initialized_dsns: set[str] = set()   # 每进程每 DSN 只建一次表


def timeframe_delta(timeframe: str) -> timedelta:
    if timeframe not in _TF_MINUTES:
        raise ValueError(f"不支持的 timeframe：{timeframe}（可选 {sorted(_TF_MINUTES)}）")
    return timedelta(minutes=_TF_MINUTES[timeframe])


class OhlcvArchive:
    def __init__(self, dsn: str | None = None) -> None:
        self._dsn = dsn or default_db_url()

    async def _conn(self) -> asyncpg.Connection:
        conn = await asyncpg.connect(self._dsn)
        if self._dsn not in _initialized_dsns:
            for stmt in _SCHEMA_STMTS:
                await conn.execute(stmt)
            _initialized_dsns.add(self._dsn)
        return conn

    async def load(self, venue_id: str, symbol: str, timeframe: str = "1h",
                   bars: int = 500) -> list[Candle]:
        """读缓存中最近 bars 根 K 线（按时间升序）。"""
        conn = await self._conn()
        try:
            rows = await conn.fetch(
                "SELECT ts, open, high, low, close, volume FROM ohlcv "
                "WHERE venue=$1 AND symbol=$2 AND timeframe=$3 ORDER BY ts DESC LIMIT $4",
                venue_id, symbol, timeframe, bars)
        finally:
            await conn.close()
        return [self._to_candle(r) for r in reversed(rows)]

    async def ensure(self, venue, symbol: str, timeframe: str = "1h",
                     bars: int = 500) -> list[Candle]:
        """确保缓存至少有 bars 根：不足则从 venue 分页拉取补齐，再返回最近 bars 根。"""
        venue_id = getattr(venue, "id", "cex")
        cached = await self.load(venue_id, symbol, timeframe, bars)
        if len(cached) >= bars:
            return cached[-bars:]

        step = timeframe_delta(timeframe)
        now = datetime.now(timezone.utc)
        since = now - step * (bars + 2)   # +2 根缓冲：对齐 K 线边界，防非整点时刻少取
        if cached:   # 已有一段历史：从最后一根之后增量补
            since = min(since, cached[-1].ts + step)

        fetched: dict[datetime, Candle] = {c.ts: c for c in cached}
        while len(fetched) < bars * 2:   # 上限护栏，防交易所返回异常数据死循环
            batch = await venue.fetch_ohlcv(symbol, timeframe=timeframe,
                                            limit=min(500, bars), since=since)
            if not batch:
                break
            new = [c for c in batch if c.ts not in fetched]
            for c in batch:
                fetched[c.ts] = c
            last_ts = batch[-1].ts
            if not new or last_ts + step > now:
                break
            since = last_ts + step

        candles = sorted(fetched.values(), key=lambda c: c.ts)
        await self._save(venue_id, symbol, timeframe, candles)
        return candles[-bars:]

    async def _save(self, venue_id: str, symbol: str, timeframe: str,
                    candles: list[Candle]) -> None:
        if not candles:
            return
        conn = await self._conn()
        try:
            await conn.executemany(
                "INSERT INTO ohlcv (venue, symbol, timeframe, ts, open, high, low, close, volume) "
                "VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) "
                "ON CONFLICT (venue, symbol, timeframe, ts) DO UPDATE SET "
                "open=EXCLUDED.open, high=EXCLUDED.high, low=EXCLUDED.low, "
                "close=EXCLUDED.close, volume=EXCLUDED.volume",
                [(venue_id, symbol, timeframe, c.ts,
                  c.open, c.high, c.low, c.close, c.volume) for c in candles])
        finally:
            await conn.close()

    @staticmethod
    def _to_candle(row) -> Candle:
        ts, o, hi, lo, c, v = row
        return Candle(ts=ts, open=Decimal(o), high=Decimal(hi), low=Decimal(lo),
                      close=Decimal(c), volume=Decimal(v))

File: internal/ohlcv/archive.go
// Package ohlcv provides durable, validated market-candle archival without
// putting PostgreSQL on the latency-sensitive trading path.
package ohlcv

import (
	"context"
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/wangchaozhi/cyp-agent/internal/contracts"
)

var timeframeDurations = map[string]time.Duration{
	"1m": time.Minute, "5m": 5 * time.Minute, "15m": 15 * time.Minute,
	"30m": 30 * time.Minute, "1h": time.Hour, "4h": 4 * time.Hour,
	"1d": 24 * time.Hour,
}

func TimeframeDuration(timeframe string) (time.Duration, error) {
	duration, ok := timeframeDurations[strings.ToLower(strings.TrimSpace(timeframe))]
	if !ok {
		return 0, fmt.Errorf("unsupported timeframe %q", timeframe)
	}
	return duration, nil
}

// ValidatedClosedCandles rejects forming, malformed and duplicate bars. A
// timestamp is interpreted as the opening time used by Binance and OKX.
func ValidatedClosedCandles(
	timeframe string,
	candles []contracts.Candle,
	now time.Time,
) ([]contracts.Candle, error) {
	duration, err := TimeframeDuration(timeframe)
	if err != nil {
		return nil, err
	}
	now = now.UTC()
	unique := make(map[int64]contracts.Candle, len(candles))
	for _, candle := range candles {
		candle.TS = candle.TS.UTC()
		if !validCandle(candle) || candle.TS.Add(duration).After(now) {
			continue
		}
		unique[candle.TS.UnixNano()] = candle
	}
	result := make([]contracts.Candle, 0, len(unique))
	for _, candle := range unique {
		result = append(result, candle)
	}
	sort.Slice(result, func(i, j int) bool { return result[i].TS.Before(result[j].TS) })
	return result, nil
}

func validCandle(candle contracts.Candle) bool {
	if candle.TS.IsZero() || !candle.Open.IsPositive() || !candle.High.IsPositive() ||
		!candle.Low.IsPositive() || !candle.Close.IsPositive() || candle.Volume.IsNegative() {
		return false
	}
	if candle.Low.Cmp(candle.High) > 0 {
		return false
	}
	return candle.Open.Cmp(candle.Low) >= 0 && candle.Open.Cmp(candle.High) <= 0 &&
		candle.Close.Cmp(candle.Low) >= 0 && candle.Close.Cmp(candle.High) <= 0
}

type HistoricalVenue interface {
	ID() string
	FetchOHLCV(context.Context, string, string, int) ([]contracts.Candle, error)
}

type RangeHistoricalVenue interface {
	HistoricalVenue
	FetchOHLCVRange(context.Context, string, string, time.Time, time.Time, int) ([]contracts.Candle, error)
}

type Archive interface {
	Load(context.Context, string, string, string, int) ([]contracts.Candle, error)
	Ensure(context.Context, HistoricalVenue, string, string, int) ([]contracts.Candle, error)
	Repair(context.Context, HistoricalVenue, string, string, time.Duration, time.Time) (int, error)
	Save(context.Context, string, string, string, []contracts.Candle) error
	Prune(context.Context, time.Time) (int64, error)
	Close()
}

type timeRange struct{ start, end time.Time }

// Repair finds missing bars in the retained window and fills each range using
// exchange pagination. It is safe to rerun because Save uses an upsert key.
func (archive *PostgresArchive) Repair(
	ctx context.Context,
	market HistoricalVenue,
	symbol, timeframe string,
	retention time.Duration,
	now time.Time,
) (int, error) {
	if market == nil {
		return 0, errors.New("historical venue is required")
	}
	duration, err := TimeframeDuration(timeframe)
	if err != nil {
		return 0, err
	}
	if retention <= 0 {
		return 0, errors.New("OHLCV repair retention must be positive")
	}
	end := now.UTC().Truncate(duration)
	start := end.Add(-retention)
	maxBars := int(retention/duration) + 2
	cached, err := archive.Load(ctx, market.ID(), symbol, timeframe, maxBars)
	if err != nil {
		return 0, err
	}
	ranges := missingTimeRanges(cached, start, end, duration)
	if len(ranges) == 0 {
		return 0, nil
	}
	ranged, supportsRange := market.(RangeHistoricalVenue)
	if !supportsRange {
		latest, fetchErr := market.FetchOHLCV(ctx, symbol, timeframe, minInt(maxBars, 200))
		if fetchErr != nil {
			return 0, fmt.Errorf("repair OHLCV history: %w", fetchErr)
		}
		closed, validationErr := ValidatedClosedCandles(timeframe, latest, now)
		if validationErr != nil {
			return 0, validationErr
		}
		if err := archive.Save(ctx, market.ID(), symbol, timeframe, closed); err != nil {
			return 0, err
		}
		return len(closed), nil
	}
	repaired := 0
	for _, missing := range ranges {
		candles, fetchErr := ranged.FetchOHLCVRange(ctx, symbol, timeframe, missing.start, missing.end, 300)
		if fetchErr != nil {
			return repaired, fmt.Errorf("repair OHLCV range %s..%s: %w",
				missing.start.Format(time.RFC3339), missing.end.Format(time.RFC3339), fetchErr)
		}
		closed, validationErr := ValidatedClosedCandles(timeframe, candles, now)
		if validationErr != nil {
			return repaired, validationErr
		}
		if err := archive.Save(ctx, market.ID(), symbol, timeframe, closed); err != nil {
			return repaired, err
		}
		repaired += len(closed)
	}
	return repaired, nil
}

func missingTimeRanges(candles []contracts.Candle, start, end time.Time, duration time.Duration) []timeRange {
	if !start.Before(end) {
		return nil
	}
	if len(candles) == 0 {
		return []timeRange{{start: start, end: end}}
	}
	sorted := append([]contracts.Candle(nil), candles...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].TS.Before(sorted[j].TS) })
	result := make([]timeRange, 0)
	cursor := start
	for _, candle := range sorted {
		timestamp := candle.TS.UTC()
		if timestamp.Before(start) || !timestamp.Before(end) {
			continue
		}
		if timestamp.After(cursor) {
			result = append(result, timeRange{start: cursor, end: timestamp})
		}
		if next := timestamp.Add(duration); next.After(cursor) {
			cursor = next
		}
	}
	if cursor.Before(end) {
		result = append(result, timeRange{start: cursor, end: end})
	}
	return result
}

func minInt(left, right int) int {
	if left < right {
		return left
	}
	return right
}

type PostgresArchive struct{ pool *pgxpool.Pool }

func NewPostgresArchive(ctx context.Context, dsn string) (*PostgresArchive, error) {
	if ctx == nil {
		return nil, errors.New("context is required")
	}
	if strings.TrimSpace(dsn) == "" {
		return nil, errors.New("PostgreSQL DSN is required")
	}
	pool, err := pgxpool.New(ctx, dsn)
	if err != nil {
		return nil, fmt.Errorf("open OHLCV archive: %w", err)
	}
	if err := pool.Ping(ctx); err != nil {
		pool.Close()
		return nil, fmt.Errorf("ping OHLCV archive: %w", err)
	}
	archive := &PostgresArchive{pool: pool}
	statements := []string{
		`CREATE TABLE IF NOT EXISTS ohlcv (
            venue TEXT NOT NULL,
            symbol TEXT NOT NULL,
            timeframe TEXT NOT NULL,
            ts TIMESTAMPTZ NOT NULL,
            open NUMERIC NOT NULL,
            high NUMERIC NOT NULL,
            low NUMERIC NOT NULL,
            close NUMERIC NOT NULL,
            volume NUMERIC NOT NULL,
            ingested_at TIMESTAMPTZ NOT NULL DEFAULT now(),
            quality_status TEXT NOT NULL DEFAULT 'validated',
            PRIMARY KEY (venue, symbol, timeframe, ts)
        )`,
		`ALTER TABLE ohlcv ADD COLUMN IF NOT EXISTS ingested_at TIMESTAMPTZ NOT NULL DEFAULT now()`,
		`ALTER TABLE ohlcv ADD COLUMN IF NOT EXISTS quality_status TEXT NOT NULL DEFAULT 'validated'`,
		`CREATE INDEX IF NOT EXISTS ohlcv_symbol_timeframe_ts_idx
            ON ohlcv (symbol, timeframe, ts DESC)`,
	}
	for _, statement := range statements {
		if _, err := pool.Exec(ctx, statement); err != nil {
			pool.Close()
			return nil, fmt.Errorf("initialize OHLCV archive: %w", err)
		}
	}
	return archive, nil
}

func (archive *PostgresArchive) Load(
	ctx context.Context,
	venueID, symbol, timeframe string,
	bars int,
) ([]contracts.Candle, error) {
	if archive == nil || archive.pool == nil {
		return nil, errors.New("OHLCV archive is closed")
	}
	if bars <= 0 {
		return []contracts.Candle{}, nil
	}
	if _, err := TimeframeDuration(timeframe); err != nil {
		return nil, err
	}
	rows, err := archive.pool.Query(ctx, `
        SELECT ts, open::text, high::text, low::text, close::text, volume::text
        FROM ohlcv
        WHERE venue=$1 AND symbol=$2 AND timeframe=$3 AND quality_status='validated'
        ORDER BY ts DESC LIMIT $4`, strings.TrimSpace(venueID), strings.TrimSpace(symbol),
		strings.ToLower(strings.TrimSpace(timeframe)), bars)
	if err != nil {
		return nil, fmt.Errorf("load OHLCV archive: %w", err)
	}
	defer rows.Close()
	result := make([]contracts.Candle, 0, bars)
	for rows.Next() {
		var timestamp time.Time
		var values [5]string
		if err := rows.Scan(&timestamp, &values[0], &values[1], &values[2], &values[3], &values[4]); err != nil {
			return nil, fmt.Errorf("scan OHLCV archive: %w", err)
		}
		parsed := [5]contracts.Decimal{}
		for index, value := range values {
			decimal, err := contracts.ParseDecimal(value)
			if err != nil {
				return nil, fmt.Errorf("decode OHLCV decimal: %w", err)
			}
			parsed[index] = decimal
		}
		result = append(result, contracts.Candle{
			TS: timestamp.UTC(), Open: parsed[0], High: parsed[1], Low: parsed[2],
			Close: parsed[3], Volume: parsed[4],
		})
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate OHLCV archive: %w", err)
	}
	for left, right := 0, len(result)-1; left < right; left, right = left+1, right-1 {
		result[left], result[right] = result[right], result[left]
	}
	return result, nil
}

// Ensure always refreshes from the venue, merges with PostgreSQL and returns
// only validated closed bars. When the venue is temporarily unavailable, a
// sufficiently deep archive remains usable for research.
func (archive *PostgresArchive) Ensure(
	ctx context.Context,
	market HistoricalVenue,
	symbol, timeframe string,
	bars int,
) ([]contracts.Candle, error) {
	if market == nil {
		return nil, errors.New("historical venue is required")
	}
	cached, err := archive.Load(ctx, market.ID(), symbol, timeframe, bars)
	if err != nil {
		return nil, err
	}
	now := time.Now()
	var fetched []contracts.Candle
	var fetchErr error
	if ranged, ok := market.(RangeHistoricalVenue); ok {
		duration, durationErr := TimeframeDuration(timeframe)
		if durationErr != nil {
			return nil, durationErr
		}
		end := now.UTC().Truncate(duration)
		start := end.Add(-time.Duration(bars) * duration)
		fetched, fetchErr = ranged.FetchOHLCVRange(ctx, symbol, timeframe, start, end, 300)
	} else {
		fetched, fetchErr = market.FetchOHLCV(ctx, symbol, timeframe, minInt(bars, 200))
	}
	if fetchErr != nil {
		if len(cached) >= bars {
			return cached[len(cached)-bars:], nil
		}
		return nil, fmt.Errorf("fetch OHLCV history: %w", fetchErr)
	}
	closed, err := ValidatedClosedCandles(timeframe, fetched, now)
	if err != nil {
		return nil, err
	}
	merged := make(map[int64]contracts.Candle, len(cached)+len(closed))
	for _, candle := range append(cached, closed...) {
		merged[candle.TS.UTC().UnixNano()] = candle
	}
	result := make([]contracts.Candle, 0, len(merged))
	for _, candle := range merged {
		result = append(result, candle)
	}
	sort.Slice(result, func(i, j int) bool { return result[i].TS.Before(result[j].TS) })
	if err := archive.Save(ctx, market.ID(), symbol, timeframe, result); err != nil {
		return nil, err
	}
	if len(result) > bars {
		result = result[len(result)-bars:]
	}
	return result, nil
}

func (archive *PostgresArchive) Save(
	ctx context.Context,
	venueID, symbol, timeframe string,
	candles []contracts.Candle,
) error {
	if archive == nil || archive.pool == nil {
		return errors.New("OHLCV archive is closed")
	}
	venueID, symbol = strings.TrimSpace(venueID), strings.TrimSpace(symbol)
	if venueID == "" || symbol == "" {
		return errors.New("OHLCV venue and symbol are required")
	}
	timeframe = strings.ToLower(strings.TrimSpace(timeframe))
	closed, err := ValidatedClosedCandles(timeframe, candles, time.Now())
	if err != nil {
		return err
	}
	if len(closed) == 0 {
		return nil
	}
	batch := &pgx.Batch{}
	for _, candle := range closed {
		batch.Queue(`INSERT INTO ohlcv
            (venue, symbol, timeframe, ts, open, high, low, close, volume, ingested_at, quality_status)
            VALUES ($1,$2,$3,$4,$5::numeric,$6::numeric,$7::numeric,$8::numeric,$9::numeric,now(),'validated')
            ON CONFLICT (venue, symbol, timeframe, ts) DO UPDATE SET
            open=EXCLUDED.open, high=EXCLUDED.high, low=EXCLUDED.low,
            close=EXCLUDED.close, volume=EXCLUDED.volume,
            ingested_at=now(), quality_status='validated'`,
			venueID, symbol, timeframe, candle.TS.UTC(), candle.Open.String(), candle.High.String(),
			candle.Low.String(), candle.Close.String(), candle.Volume.String())
	}
	results := archive.pool.SendBatch(ctx, batch)
	for range closed {
		if _, err := results.Exec(); err != nil {
			_ = results.Close()
			return fmt.Errorf("save OHLCV archive: %w", err)
		}
	}
	if err := results.Close(); err != nil {
		return fmt.Errorf("close OHLCV batch: %w", err)
	}
	return nil
}

func (archive *PostgresArchive) Prune(ctx context.Context, before time.Time) (int64, error) {
	if archive == nil || archive.pool == nil {
		return 0, errors.New("OHLCV archive is closed")
	}
	result, err := archive.pool.Exec(ctx, `DELETE FROM ohlcv WHERE ts < $1`, before.UTC())
	if err != nil {
		return 0, fmt.Errorf("prune OHLCV archive: %w", err)
	}
	return result.RowsAffected(), nil
}

func (archive *PostgresArchive) Close() {
	if archive != nil && archive.pool != nil {
		archive.pool.Close()
	}
}

var _ Archive = (*PostgresArchive)(nil)
